[torch/elastic] Introduce `C10dRendezvousBackend`.

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55637 [torch/elastic] Introduce `EtcdRendezvousBackend`.
* **#55636 [torch/elastic] Introduce `C10dRendezvousBackend`.**
* #55635 [torch/elastic] Introduce `DynamicRendezvousHandler` and `RendezvousBackend`.

This diff introduces:

- The `C10dRendezvousBackend` type to support C10d stores as rendezvous backends.
- A fix to the `TCPStore.compare_set()` function to support non-existent keys.
- A placeholder `c10d-experimental` registry to instantiate C10d-baked rendezvous backends via `get_rendezvous_handler()`.

Differential Revision: [D27654492](https://our.internmc.facebook.com/intern/diff/D27654492/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654492/)!

cbalioglu committed:
[torch/elastic] Introduce `C10dRendezvousBackend`.

This diff introduces:

- The `C10dRendezvousBackend` type to support C10d stores as rendezvous backends.
- A fix to the `TCPStore.compare_set()` function to support non-existent keys.
- A placeholder `c10d-experimental` registry to instantiate C10d-baked rendezvous backends via `get_rendezvous_handler()`.

Differential Revision: [D27654492](https://our.internmc.facebook.com/intern/diff/D27654492/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654492/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+        read_timeout (int, optional):
+            The read timeout, in seconds, for store operations. Defaults to 60
+            seconds.
+        is_host (bool, optional):
```
Probably better not have a TCPStore specific flag like this. Maybe we can have an argument like `**store_options` which can be used to specific store specific options?

cbalioglu replied:
All these "parameters" (`store_type`, `read_timeout`, and `is_host`) are in fact configuration options passed as `**kwargs` to `RendezvousParameters` and read from it using its generic `get()` method. See [here](https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/api.py#L160) and [here](https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/api.py#L182).

`RendezvousParameters` is a backend-agnostic type that has typed attributes only for mandatory parameters such as minimum and maximum number of nodes, and the rendezvous endpoint. Any backend specific configuration is passed via `**kwargs`. The `create_handler()` function is then responsible for parsing its own specific arguments from those `kwargs`.

As I mentioned above `create_handler()` is meant to instantiate a backend from command line arguments which are passed indirectly via `RendezvousParameters`. If the API is used programmatically, one should construct the backend using its `__init__`.

Maybe I should make `create_handler()` internal (meaning `_create_handler()`) to make it more clear that it is not intended to be called in a normal API usage.

pritamdamania87 commented on test/distributed/elastic/rendezvous/c10d_rendezvous_backend_test.py:
```diff
@@ -0,0 +1,155 @@
+
+    def test_create_backend_returns_backend_if_is_host_is_false(self) -> None:
+        store = TCPStore(  # type: ignore[call-arg] # noqa: F841
+            self._expected_endpoint_host, self._expected_endpoint_port, is_master=True
+        )
+
+        self._params.config["is_host"] = "false"
```
I'm a bit confused by this test, we end up creating the TCPStore twice here? Once on line 72 and then again in `test_create_backend_returns_backend` as part of create_backend? What exactly are we testing here that is not already tested by test_create_backend_returns_backend? Is it the fact that we can initialize two TCPStores on different ports?

pritamdamania87 replied:
Also, how are we able to create two instances of TCPStore on the same port?

cbalioglu replied:
> Also, how are we able to create two instances of TCPStore on the same port?

That is in fact what we are testing here. It simulates the case where more than one process on the same machine is part of the same job. Although quite an edge case in production, this can happen in test and development environments. Say for instance you instantiate a job with two workers on a single machine for testing purposes:

```
$ torch/distributed/elastic-launch --rendezvous-backend=c10d --rendezvous-endpoint=localhost:1234 ...
$ torch/distributed/elastic-launch --rendezvous-backend=c10d --rendezvous-endpoint=localhost:1234 ...
```

Because our endpoint is "localhost" both processes will attempt to instantiate the `TCPStore` with `is_master=True`. Obviously only one of them will be able to host the `TCPStore` and the other one will fail with something like "socket already in use". That second process will then re-attempt to instantiate the `TCPStore`, but this time it will explicitly pass `is_master=False`.

Note that this is the default behavior which would work in most cases. However if, for some reason, you want to override which exact process should be the store host, you can explicitly pass a `is_host` configuration option:

```
$ torch/distributed/elastic-launch --rendezvous-backend=c10d --rendezvous-endpoint=localhost:1234 --args is_host=false ...
$ torch/distributed/elastic-launch --rendezvous-backend=c10d --rendezvous-endpoint=localhost:1234 --args is_host=true ...
```

Now you know for sure that the second process will be the host of the store.

pritamdamania87 commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+    """
+
+    # See the explanation in the __init__ method.
+    _NULL_SENTINEL = "Y2FuaW1hZGFt"
```
Can we use something like "<null>" for this instead or is it likely the rendezvous state might consist of something like this?

cbalioglu replied:
I wanted to have something that is practically unique. Actually `Y2FuaW1hZGFt` means "my dear Ada" (my daughter's name) in Turkish in base64-encoded form :) So very unlikely that it will match a binary string produced by Python's pickle. I thought about using an empty string, but I wasn't sure whether it would be always an invalid state value (what if we optimize our state serialization logic in the future and use "" for initial state). Instead of taking the risk, I just came up with this "random" string.

pritamdamania87 commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+        else:
+            token = self._NULL_SENTINEL
+
+        base64_state: bytes = self._call_store("compare_set", self._key, token, base64_state_str)
```
Why do we have a call_store API like this instead of just using `self._store.compare_set(....)`?

cbalioglu replied:
We call the store in several places in `C10dRendezvousBackend`. I wanted to make sure that we wrap each call with an identical try/except clause to propagate errors consistently as `RendezvousConnectionError`s (where the original error is put into `__cause__`). I could either duplicate the same try/except in all places where we call the store, or come up with a fancy decorator or context manager to "magically" wrap them, but thought that this would be the most straightforward solution.

pritamdamania87 commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+    # Otherwise try to determine whether we are the host based on our hostname
+    # and IP address.
+    else:
+        is_host = _matches_machine_hostname(host)
```
Would it be simpler if just required `is_host` to be always specified and error out if cfg_is_host is None?

cbalioglu replied:
It would be definitely simpler on our end, but it would also be slightly inconvenient and redundant for the user. In normal circumstances (except for the edge case I mentioned above where you might have more than one process on the same machine) the endpoint already conveys all the information you need to determine whether you are the host or not. If the specified endpoint's hostname equals your machine's hostname, you know that you will be the host. Also this way you can have a single launcher script that can be used by all machines identically. If `is_host` would be mandatory, you would have to make sure that there are two versions of the very same launcher script; one for the machine that corresponds to the endpoint with `is_host=True`, and one for other machines with `is_host=False`.

pritamdamania87 commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+    for _ in range(2):
+        try:
+            store = TCPStore(  # type: ignore[call-arg]
+                host, port, is_master=is_host, timeout=timedelta(seconds=read_timeout)
```
Don't we need the ability to pass the world_size here as well?

cbalioglu replied:
world_size is determined after a round of rendezvous is complete. This `TCPStore` instance is only used by the TorchElastic agents for synchronization between them. The "actual" store used by workers is later handled using a different mechanism inside an agent.

pritamdamania87 commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+                    "The connection to the C10d store has failed. See inner exception for details."
+                ) from exc
+
+            is_host = False
```
Wouldn't this be retried by all processes in the world_size with `is_host=False` leading to all of them failing again since no one has `is_host=True`?

cbalioglu replied:
There are three cases:

- If `is_host` was explicitly provided by the user, we do not re-attempt. After the first attempt, if the store initialization fails, we simply raise an error.
- If we inferred the value of `is_host` to be `False` (our hostname does not match the hostname of the specified rendezvous endpoint), again; we do not re-attempt and raise an error.
- If we inferred the value of `is_host` to be `True` (our hostname matches the hostname of the specified rendezvous endpoint), only then we attempt a second time with `is_master=False` assuming another process on the same machine is also part of the same job and acted faster than us.

pritamdamania87 commented on torch/distributed/elastic/rendezvous/registry.py:
```diff
@@ -14,8 +15,17 @@ def _create_etcd_handler(params: RendezvousParameters) -> RendezvousHandler:
+
 def _register_default_handlers() -> None:
     handler_registry.register("etcd", _create_etcd_handler)
+    handler_registry.register("c10d-experimental", _create_expr_c10d_handler)
```
What is the reason to name this `c10d-experimental` instead of just `c10d`? We usually tag things as "experimental" in the docs instead.

cbalioglu replied:
Well I thought we should clearly indicate to the user that this is still a fairly new feature that might not be stable yet. If there is already a convention to convey such information solely in the docs, I can clearly change it to `c10d`. Will do that in the next revision.

pritamdamania87 commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+
+            break
+        except (ValueError, RuntimeError) as exc:
+            # If we heuristically inferred the value of is_host as True and our
+            # first attempt to instantiate the TCP store has failed, try it one
+            # more time with is_host set to False. As an edge case there can be
+            # more than one process that is part of the same rendezvous on this
+            # machine and only one of them will eventually host the store.
```
Would this be much simpler as well if we require `is_host` to be specified?

cbalioglu replied:
Yes, please see my answer above for line 127.

cbalioglu commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+    """Creates a new `C10dRendezvousBackend` from the specified parameters.
+
+    Configuration:
+        store_type (str, optional):
```
These configuration options are stored in a `Dict[str, str]` inside the `RendezvousParameters` type. The `RendezvousParameters` instance passed to `create_backend()` is constructed by the launcher script and configuration options are read from the command line without any special interpretation (see [here](https://github.com/pytorch/elastic/blob/master/torchelastic/distributed/launch.py#L464) and [here](https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/utils.py#L11)). The type annotations `str` and `int` you see here are more about what those command line strings should represent, rather than their actual Python type.

It is the job of `create_handler()` to parse the required parameters from `RendezvousParameters` and then pass them as typed arguments to the constructor of `C10dRendezvousBackend`. You can think of `create_handler()` as a helper function to instantiate a `C10dRendezvousBackend` from command line arguments.

The `create_handler()` function here follows the idiom that has been used by the existing `etcd` and `zeus` rendezvous handlers by the way. Each backend has its corresponding `create_handler()` function which is used to delegate the instantiation of that particular backend. See [`registry.py`](https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/registry.py) as an example.

cbalioglu commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+        store_type (str, optional):
+            The type of the C10d store. As of today the only supported type is
+            "tcp" which corresponds to `TCPStore`. Defaults to "tcp".
+        read_timeout (int, optional):
```
See my answer below for `store_type`.

H-Huang commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+        #
+        # As a workaround we initially set a sentinel value as the rendezvous
+        # state. Whenever this value gets returned we treat it as a None.
+        self._call_store("compare_set", self._key, "", self._NULL_SENTINEL)
```
Agree with changing to "set" as @tierex mentioned. Then we also wouldn't need to change compareSetHandler in TCPStore.cpp

cbalioglu replied:
Please see my answer to him in Phabricator.

H-Huang commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+            the FQDN of the machine).
+    """
+    # As of today we only support TCPStore. Other store types do not have the
+    # required functionality (e.g. compare_set) yet.
```
Just FYI, compare_set was added to all stores. So FileStore should work if all workers have access to the same file, no need to include support for FileStore for this PR though.

H-Huang commented on torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py:
```diff
@@ -0,0 +1,195 @@
+    if store_type != "tcp":
+        raise ValueError("The store type must be 'tcp'. Other store types are not supported yet.")
+
+    store = _create_tcp_store(params)
```
Is there a place where we log the RendezvousBackend being created and what params it uses? That would be useful for debugging

cbalioglu replied:
Once we finish the initial round of reviews, I can certainly add more logging statements.

cbalioglu committed:
Update on "[torch/elastic] Introduce `C10dRendezvousBackend`."


This diff introduces:

- The `C10dRendezvousBackend` type to support C10d stores as rendezvous backends.
- A fix to the `TCPStore.compare_set()` function to support non-existent keys.
- A placeholder `c10d-experimental` registry to instantiate C10d-baked rendezvous backends via `get_rendezvous_handler()`.

Differential Revision: [D27654492](https://our.internmc.facebook.com/intern/diff/D27654492/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654492/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "[torch/elastic] Introduce `C10dRendezvousBackend`."


This diff introduces:

- The `C10dRendezvousBackend` type to support C10d stores as rendezvous backends.
- A fix to the `TCPStore.compare_set()` function to support non-existent keys.
- A placeholder `c10d-experimental` registry to instantiate C10d-baked rendezvous backends via `get_rendezvous_handler()`.

Differential Revision: [D27654492](https://our.internmc.facebook.com/intern/diff/D27654492/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654492/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "[torch/elastic] Introduce `C10dRendezvousBackend`."


This diff introduces:

- The `C10dRendezvousBackend` type to support C10d stores as rendezvous backends.
- A fix to the `TCPStore.compare_set()` function to support non-existent keys.
- A placeholder `c10d-experimental` registry to instantiate C10d-baked rendezvous backends via `get_rendezvous_handler()`.

Differential Revision: [D27654492](https://our.internmc.facebook.com/intern/diff/D27654492/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27654492/)!

[ghstack-poisoned]

rohan-varma commented on test/distributed/elastic/rendezvous/c10d_rendezvous_backend_test.py:
```diff
@@ -0,0 +1,155 @@
+from typing import ClassVar
+from unittest import TestCase
+
+from torch.distributed import TCPStore
```
A bit late to this PR, but is the goal of `C10dRendezvousBackend` to be able to support other stores such as FileStore and HashStore as well, or only TCPStore?

cbalioglu replied:
As of today it only supports `TCPStore` (read it as "only tested with"), but it is implemented in a generic way and can potentially be used with any `Store` type that has a `compare_set` function. In fact Howard recently implemented `compare_set` for `FileStore`. This means technically we can also use it with `FileStore`.
